Add first version of bootstrap config support

geoand opened the pull request:
This implements the design that was discussed in https://github.com/smallrye/smallrye-config/issues/215.

It introduces a new `Bootstrap` configuration phase that can be consumed by a Recorder whose intent is to produce a `ConfigSourceProvider`. These recorders are run during runtime init, but before the other Recorders. Their result contributes to the final runtime `Config` creation.

The part I am not too excited about is the integration into the build system since the new build items don't utilize the full power of the build system. However, 1) appropriate error messages are provided so people looking to create a bootstrap config implementation will be properly guided, 2) I don't see a good way of making it better (and avoiding hard-coding the specific Build Items) unless we introduce a new recording phase as well.

geoand committed:
Add first version of bootstrap config support

machi1990 commented on core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigPhase.java:
```diff
@@ -6,6 +6,8 @@
 public enum ConfigPhase implements Comparable<ConfigPhase> {
     RUN_TIME("The configuration is overridable at runtime", Constants.CONFIG_PHASE_RUNTIME_ILLUSTRATION, "RunTime"),
+    BOOTSTRAP("The configuration is used to bootstrap runtime Config Sources and is overridable at runtime",
+            Constants.CONFIG_PHASE_BOOTSTRAP_ILLUSTRATION, "BuildTime"),
```
```suggestion
            Constants.CONFIG_PHASE_BOOTSTRAP_ILLUSTRATION, "Bootstrap"),
```

geoand replied:
Makes sense, thanks

dmlloyd commented on ci-templates/stages.yml:
```diff
@@ -346,12 +346,13 @@ stages:
           poolSettings: ${{parameters.poolSettings}}
           expectUseVMs: ${{parameters.expectUseVMs}}
-          timeoutInMinutes: 30
+          timeoutInMinutes: 35
```
Uh oh. :slightly_smiling_face:

geoand replied:
Well this was just done to be on the safe side since I added a new test :). 
I actually didn't see any actual problem :P

dmlloyd commented on core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigPhase.java:
```diff
@@ -37,6 +39,20 @@ public int compare(ConfigPhase firstPhase, ConfigPhase secondPhase) {
                             return -1;
                     }
                 }
+                case BOOTSTRAP: {
```
Heh! This bit of code probably should not have been approved - instead the phases probably ought to have been reordered and the default comparison used.... but I guess that can be a fix for another day.

geoand commented:
@dmlloyd I'll add my comments later on today or tomorrow morning.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java:
```diff
@@ -213,24 +207,61 @@ MainClassBuildItem build(List<StaticBytecodeRecorderBuildItem> staticInitTasks,
+             * the generated StartupTask will write its result (which is a RuntimeValue<ConfigSourceProvider>
+             * configSourcesValue) into the StartupContext.
+             * This value is then pulled out of the StartupContext by using the getLastValue method. All the
+             * ConfigSourceProvider objects are then collected and passed to Config.readConfig()
```
Yes certainly. I think I have identified the mental hurdle that has been blocking for me seeing the path forward with this way of doing things.
I'll work on it tomorrow and see how it goes.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -733,8 +744,14 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
+                        } else if (phase == ConfigPhase.BOOTSTRAP || phase == ConfigPhase.RUN_TIME) {
                             if (isRecorder) {
+                                if ((phase == ConfigPhase.BOOTSTRAP)
+                                        && !method.getReturnType().equals(RunTimeConfigurationSourceValueBuildItem.class)) {
```
I agree completely. The whole reason I did it this way is because I wanted to be able to distringuish the bootstrap recorders for the regular ones and therefor generate bytecode that interacts with Config properly.
If I can change the PR to leverage the Build Steps that you mention in order comments, than this and other comments on the PR will likely go away.

geoand replied:
This arbitrary restriction has now been lifted.

geoand committed:
Properly integrate bootstrap config into build system

This round makes bootstrap config work properly with
the build system without the arbitrary restrictions of
the first iteration.
This is accomplished mostly by moving config generation
and setup into their own build steps

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -634,9 +644,10 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
                 final boolean optional = recordAnnotation.optional();
+
                 methodStepConfig = methodStepConfig.andThen(bsb -> bsb.produces(
                         executionTime == ExecutionTime.STATIC_INIT ? StaticBytecodeRecorderBuildItem.class
```
This junk is now completely gone

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -778,6 +797,12 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
                 resultConsumer = Functions.discardingBiConsumer();
             } else if (rawTypeExtends(returnType, BuildItem.class)) {
                 final Class<? extends BuildItem> type = method.getReturnType().asSubclass(BuildItem.class);
+                if (type.equals(RunTimeConfigurationSourceValueBuildItem.class)
```
Also done

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -778,6 +797,12 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
+                        && (!isRecorder || recordAnnotation.value() != ExecutionTime.RUNTIME_INIT)) {
+                    throw reportError(method,
+                            "A Build step that returns " + RunTimeConfigurationSourceValueBuildItem.class.getSimpleName()
+                                    + " must also be annotated with @Record(ExecutionTime.RUNTIME_INIT)");
```
Gone as well

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -837,14 +862,20 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
+            if (methodConsumingConfigPhases.contains(ConfigPhase.BOOTSTRAP)
+                    && methodConsumingConfigPhases.contains(ConfigPhase.RUN_TIME)) {
+                throw reportError(method,
+                        "Bootstrap configuration cannot be used together in a build step with run time configuration");
```
Not needed any more

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java:
```diff
@@ -361,19 +390,29 @@ public void run() {
-                    runTimePatternMap, combinator);
-            final ConfigPatternMap<Container> runTimeIgnored = ConfigPatternMap.merge(buildTimePatternMap,
-                    buildTimeRunTimePatternMap, combinator);
+            ConfigPatternMap<Container> buildTimeRunTimeIgnored = ConfigPatternMap
```
Fixed

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java:
```diff
@@ -213,24 +207,61 @@ MainClassBuildItem build(List<StaticBytecodeRecorderBuildItem> staticInitTasks,
+             * stuff executes.
+             * It's tailored made to support the bootstrap configuration stuff and would need to be rethought if we need
+             * a general mechanism of executing things before the main bytecode stuff
+             * (which would likely involve a new runtime phase)
```
It worked like a charm

geoand commented:
@dmlloyd The solution has been updatedt to incorporate your suggestion of moving the config loading into build steps and using build items that allow the build system to work out the proper order of things.

I feel much better about this iteration - hopefully you will too :laughing:

geoand commented:
@dmlloyd kind reminder 😁

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java:
```diff
@@ -213,25 +178,60 @@ MainClassBuildItem build(List<StaticBytecodeRecorderBuildItem> staticInitTasks,
 
-        // Load the run time configuration
-        tryBlock.invokeStaticMethod(RunTimeConfigurationGenerator.C_CREATE_RUN_TIME_CONFIG);
+        //        // Load the bootstrap configuration
```
This should all go away.  Might be worth tagging the commit for reference first.

geoand replied:
Oh darn! Sorry for missing that! Basically the PR has 2 commits the first approach and the new approach so everything should be in the history, so I think we can safely delete all this :P

dmlloyd reviewed (changes requested):
Looking better, I think the general approach seems OK.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/MainBytecodeRecorderBuildItem.java:
```diff
@@ -6,13 +6,23 @@
 public final class MainBytecodeRecorderBuildItem extends MultiBuildItem {
 
     private final BytecodeRecorderImpl bytecodeRecorder;
+    private final String generatedStartupContextClassName;
```
The reason I did it this way is because the `MainClassBuildStep` can continue to just consume `List<MainBytecodeRecorderBuildItem>` and the order is properly worked out.

dmlloyd replied:
I'm still not loving this, but the Monday brain fails to come up with any better ideas, so... I guess it's fine.

geoand replied:
Same here :)

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -743,7 +753,9 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
-                                        "Run time configuration cannot be consumed here unless the method is a @Recorder");
+                                        String.format(
+                                                "%s configuration cannot be consumed here unless the method is a @Recorder",
+                                                phase == ConfigPhase.RUN_TIME ? "Run time" : "Bootstrap"));
```
Done

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java:
```diff
@@ -733,7 +743,7 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
                                 runTimeProxies.computeIfAbsent(parameterClass, readResult::requireRootObjectForClass);
                             }
-                        } else if (phase == ConfigPhase.RUN_TIME) {
+                        } else if (phase == ConfigPhase.BOOTSTRAP || phase == ConfigPhase.RUN_TIME) {
```
Done

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/BootstrapConfigSetupCompleteBuildItem.java:
```diff
@@ -0,0 +1,10 @@
+ * Marker used by Build Steps that consume bootstrap configuration to ensure that they run after the bootstrap config has been
+ * setup
+ */
+public final class BootstrapConfigSetupCompleteBuildItem extends SimpleBuildItem {
```
Done for this one and for `RuntimeConfigSetupCompleteBuildItem`

geoand commented:
@dmlloyd I think your concerns should now be addressed

geoand commented:
@dmlloyd is there anything else you think needs to be changed for this one?

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/RunTimeConfigurationSourceValueBuildItem.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * This is a special build item that is intended to be used only to support bootstrap configuration in the following manner:
+ *
+ * A build step produces this BuildItem with a {@code RuntimeValue<ConfigSourceProvider>} a payload that is obtained by calling
```
```suggestion
 * A build step produces this BuildItem with a {@code RuntimeValue&lt;ConfigSourceProvider&gt;} a payload that is obtained by calling
```

geoand commented:
@dmlloyd suggestion applied

dmlloyd commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java:
```diff
@@ -296,6 +312,10 @@ public static void generate(BuildTimeConfigurationReader.ReadResult readResult,
 
+            // static field containing the instance of the class - is set when createBootstrapConfig is run
+            cc.getFieldCreator(C_INSTANCE)
+                    .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC);
```
```suggestion
                    .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_VOLATILE);
```

geoand commented:
LOL, a suggested change caused the validator to fail, I never thought that would happen :P
